kube-client/src/discovery/apigroup.rs

```diff
@@ -133,11 +126,11 @@ impl ApiGroup {
         };
         for res in &list.resources {
             if res.kind == gvk.kind && !res.name.contains('/') {
-                let ar = parse::parse_apiresource(res, &list.group_version).map_err(
+                let mut ar = parse::parse_apiresource(res, &list.group_version).map_err(
                     |ParseGroupVersionError(s)| Error::Discovery(DiscoveryError::InvalidGroupVersion(s)),
                 )?;
-                let caps = parse::parse_apicapabilities(&list, &res.name)?;
-                return Ok((ar, caps));
+                ar.subresources = parse::find_subresources(&list, &res.name)?;
+                return Ok(ar);
```

Comment:
Would it make sense to reorganize this to:

```rust
struct ApiResourceMeta; // current ApiResource
struct ApiCapabilities; // current ApiCapabilities
struct ApiResource { // current (ApiResource, ApiCapabilities)
    meta: ApiResourceMeta,
    caps: ApiCapabilities,
}
```

That way we could provide wrapper functions as needed on `ApiResource` to avoid having the user care about the distinction, but also statically guarantee that caps are actually available when required. We might even want to `impl Deref<Target = ApiResourceMeta> for ApiResource` to approximate subtyping...

Reply:
Sorry, `s/caps/subresources/g`.

Reply:
I think that sounds like the nicest possible way to split the structs. It is still a bit awkward though:

- the [upstream struct](https://docs.rs/k8s-openapi/latest/k8s_openapi/apimachinery/pkg/apis/meta/v1/struct.APIResource.html) does not have a split like that
- we are not really able to split `APIResource` into Caps/Info now (where does short_names go? is scope a capability?), and it might be even harder in the future if upstream adds more info
- we probably need to push the distinction stuff into `ApiResourceMeta` instead as a user could want to query the api with just the GVK + plural, so now they need a partial `ApiResourceMeta` constructed somehow

The distinction here might not be that big of a deal because if we provide clean paths to get fully speced out variants of `ApiResource` via `kube-derive` or `discovery` then that's the happy path for users. If people want to construct them manually, then they probably should care about the distinction.

(As for the subresource parts of `ApiResource`, I'm not entirely sure what is the best way to represent that inside `ApiResource`. Nesting `ApiResource` inside `ApiResource` to represent subresources is unnecessary (and leaves many fields unused). Hoping to experiment and clean up that part in the client api pr - for now have left it as it was)

Reply:
What do you think?

Reply:
After looking into it a bit more it feels like there was a useful distinction where `ApiResource` was "the information you need to start querying the API" and `ApiCapabilities` was "auxilary data that might be useful for a UI (like kubectl)". I think it makes sense to keep those types distinct to document whether `ApiCapabilities` is _also_ available at some given point.

Where the old API (IMO) fell down was in that having to deal with `ApiCapabilities` ended up being a constant annoyance for people who only care about `ApiResource`.

`APIResource` is a slightly different beast, since it's always generated by the K8s API server, which always knows the canonical information for all of the fields. `from_gvk` isn't really an API that makes sense for the K8s type.

> we are not really able to split APIResource into Caps/Info now (where does short_names go? is scope a capability?), and it might be even harder in the future if upstream adds more info

`short_names` is definitely an info/cap/whatever-we-end-up-calling-it. `scope` I could see going both ways, but I'm leaning towards info as well since if you already have the object (or know how to construct it) then you can infer the scope based on that. If you want to guide the user towards how to construct the object then you'll want an `ApiResourceInfo`.

Reply:
It's not critical, but it's a thing we can and should do better, and the new client api is a chance to do that.

> The big advantage of these guard rails is being able to catch these errors at compile time, and you're already opting out of that when you use the dynamic API.

People in many cases **have** to use the dynamic api because the structs do not exist, and one of the most common errors people struggle with is people using resources in the wrong way. Having wide open apis (even if it is just for the dynamic case) is something I would like to avoid. We can only provide a safety restriction for the dynamic api at runtime, but that can still be a lot less confusing than a 404 (which makes people question if kube really works at all with their case, and then raise issues to us).

Reply:
Separating out the core bits kinda sucks, but is ~the simplest way to separate the cases on the type level. It would also make it trivial to drop the caps.

Another alternative would be to define something like:

```rust
struct ApiResource<Caps: ApiCapabilities = KnownCapabilities> {
  // core stuff
  caps: Caps,
}
struct UnknownCapabilities;
// which leads to the equivalence
type ApiResourceCore = ApiResource<UnknownCapabilities>;
```

That basically ends up hoisting the same `Option` logic into the type system. One downside of this is that `ApiResource` can't easily be downgraded into `ApiResourceCore`.

> note that i would rather not nest the core bits into its own struct because it always have to be there, and then the usability benefit of removing the pair everywhere feels somewhat reduced.

The difference is that:

1. The user doesn't have to name the whole tuple type constantly, especially when just shoving it into another kube-rs API
2. The user doesn't have to care, if both `ApiResource` and `ApiResourceCore` implement the common trait (and `Client` is defined in terms of that) then they can just pass `&ApiResource` as is

Where you do feel the distinction is if you start mucking about with the internals of `ApiResource`, but that should be a relatively niche use case IMO.

Reply:
> What do you mean? What are the use cases where you need only caps and not the core stuff? They are constructed together. You might have core + caps, but you never have just caps.

None (that I see). Either you want core or core+caps.

Embedding caps as a field on core prevents us from having distinct types for those cases. At most we could option-wrap it and distinguish it at runtime, as you did.

Reply:
Have put a `capabilities` member as `Option<ApiCapabilities>` directly on `ApiResource` now.

Have not included `namespaced` in there because `namespaced` is currently the one thing we can find through all forms of construction:

- kube-derive/discovery (exists by construction)
- ApiResource::erase (using the new Scope trait through Resource)
- ApiResource manual construction (if setting namespaced manually - although that's up to the user to get right in that case)

In fact, namespaced is now universally accessible so have made a `Resource::is_namespaced` method to help determine the namespace (which even works in the dynamic case).

Having `namespaced` within `ApiCapabilities` then we would need to partially construct it in `ApiResource::erase` and that means all fields within `ApiCapabilities` need to be option wrapped, which is not great for users of the `dynamic_api` and similar (multiple guaranteed unwraps necessary in their code). Currently it's now limited to a single unwrap in dynamic_api to access the caps:

```
// use discovery to get an AR
let caps = ar.capabilities.as_ref().expect("guaranteed because we used discovery");
caps.supports_operation(...); // can now do a check without option unwrapping from caps
```

Reply:
Have lifted the `namespaced` into capabilities struct by removing the option, and relying on default instead.

I think this is the least worst option, that still preserves a reasonable split (although the setters have been moved into `ApiResource` because we need to set `namespaced` from the root (it's the resource that can extract it) - but that's fine, these setters are only set by either kube-derive or hardcore manual users anyway).

So now ApiCapabilities has namespace/verbs/subresources/shortnames, and all of these will be set in the happy path (kube-derive/discovery). It's the happy  path we want to be as frictionless as possible, so having options for just the manual path ends up mandating unwraps on the happy path which is not great.

Reply:
I think we just keep talking past each other.

Reply:
Okay, I did a bit more digging.

I tried implementing the `ApiResource`/`ApiResourceCore` split to see how it'd go, and `Deref`/`DerefMut` helped hide it from most usage sites (aside from constructors): https://github.com/teozkr/kube-rs/tree/rm-caps-/typed.

However, this also highlit another problem with merging caps into `ApiResource`: `ApiResource` is currently embedded into `ObjectRef<DynamicObject>`, so *any* fields on it are also considered for `ObjectRef`'s `Eq`/`Hash` implementations. This means that an `ObjectRef<DynamicObject>` constructed using a discovered type and an `ObjectRef` constructed by erasing a `k8s_openapi` type would not be considered equal, even if they refer to same version of the exact same Kubernetes object!.